File: backend/src/models/MedicalPhoto.js
const { DataTypes } = require('sequelize');

module.exports = (sequelize) => {
  const MedicalPhoto = sequelize.define('MedicalPhoto', {
    id: {
      type: DataTypes.UUID,
      defaultValue: DataTypes.UUIDV4,
      primaryKey: true
    },
    certificateId: {
      type: DataTypes.UUID,
      allowNull: false,
      references: {
        model: 'MedicalCertificates',
        key: 'id'
      }
    },
    userId: {
      type: DataTypes.UUID,
      allowNull: false,
      references: {
        model: 'Users',
        key: 'id'
      }
    },
    requestedById: {
      type: DataTypes.UUID,
      allowNull: false,
      references: {
        model: 'Users',
        key: 'id'
      },
      comment: 'Médico que solicitó la foto'
    },
    // Información de la foto
    photoUrl: {
      type: DataTypes.STRING,
      allowNull: false
    },
    thumbnailUrl: {
      type: DataTypes.STRING,
      allowNull: true
    },
    originalFileName: {
      type: DataTypes.STRING,
      allowNull: true
    },
    fileSize: {
      type: DataTypes.INTEGER,
      allowNull: true
    },
    // Detalles médicos
    bodyPart: {
      type: DataTypes.STRING,
      allowNull: false,
      comment: 'Parte del cuerpo fotografiada'
    },
    bodyPartDetail: {
      type: DataTypes.STRING,
      allowNull: true,
      comment: 'Detalle específico (ej: "mano izquierda", "rodilla derecha")'
    },
    photoType: {
      type: DataTypes.ENUM('injury', 'lesion', 'swelling', 'rash', 'wound', 'other'),
      allowNull: false,
      comment: 'Tipo de afección fotografiada'
    },
    description: {
      type: DataTypes.TEXT,
      allowNull: true,
      comment: 'Descripción de lo que se ve en la foto'
    },
    // Solicitud del médico
    requestReason: {
      type: DataTypes.TEXT,
      allowNull: false,
      comment: 'Motivo por el cual el médico solicita la foto'
    },
    requestInstructions: {
      type: DataTypes.TEXT,
      allowNull: true,
      comment: 'Instrucciones específicas para tomar la foto'
    },
    requestDate: {
      type: DataTypes.DATE,
      allowNull: false,
      defaultValue: DataTypes.NOW
    },
    // Respuesta del empleado
    employeeNotes: {
      type: DataTypes.TEXT,
      allowNull: true,
      comment: 'Notas adicionales del empleado sobre la foto'
    },
    photoDate: {
      type: DataTypes.DATE,
      allowNull: true,
      comment: 'Fecha cuando se tomó la foto'
    },
    // Estado
    status: {
      type: DataTypes.ENUM('requested', 'uploaded', 'reviewed', 'rejected'),
      defaultValue: 'requested'
    },
    isRequired: {
      type: DataTypes.BOOLEAN,
      defaultValue: true,
      comment: 'Si es obligatoria para continuar el proceso'
    },
    // Revisión médica de la foto
    medicalReview: {
      type: DataTypes.TEXT,
      allowNull: true,
      comment: 'Comentarios del médico sobre la foto'
    },
    reviewedAt: {
      type: DataTypes.DATE,
      allowNull: true
    },
    reviewedById: {
      type: DataTypes.UUID,
      allowNull: true,
      references: {
        model: 'Users',
        key: 'id'
      }
    },
    // Metadata
    metadata: {
      type: DataTypes.JSON,
      allowNull: true,
      comment: 'Datos adicionales de la imagen (resolución, geolocalización, etc.)'
    }
  }, {
    tableName: 'medical_photos',
    timestamps: true,
    indexes: [
      {
        fields: ['certificateId']
      },
      {
        fields: ['userId']
      },
      {
        fields: ['requestedById']
      },
      {
        fields: ['status']
      },
      {
        fields: ['bodyPart']
      },
      {
        fields: ['photoType']
      }
    ]
  });

  // Asociaciones
  MedicalPhoto.associate = (models) => {
    // Una foto pertenece a un certificado médico
    MedicalPhoto.belongsTo(models.MedicalCertificate, {
      foreignKey: 'certificateId',
      as: 'certificate'
    });

    // Una foto pertenece a un empleado
    MedicalPhoto.belongsTo(models.User, {
      foreignKey: 'userId',
      as: 'employee'
    });

    // Una foto fue solicitada por un médico
    MedicalPhoto.belongsTo(models.User, {
      foreignKey: 'requestedById',
      as: 'requestedBy'
    });

    // Una foto puede ser revisada por un médico
    MedicalPhoto.belongsTo(models.User, {
      foreignKey: 'reviewedById',
      as: 'reviewedBy'
    });
  };

  return MedicalPhoto;
};

File: frontend_flutter/lib/models/medical_photo.dart
class MedicalPhoto {
  final String id;
  final String certificateId;
  final String userId;
  final String requestedById;
  
  // Información de la foto
  final String? photoUrl;
  final String? thumbnailUrl;
  final String? originalFileName;
  final int? fileSize;
  
  // Detalles médicos
  final String bodyPart;
  final String? bodyPartDetail;
  final String photoType;
  final String? description;
  
  // Solicitud del médico
  final String requestReason;
  final String? requestInstructions;
  final DateTime requestDate;
  
  // Respuesta del empleado
  final String? employeeNotes;
  final DateTime? photoDate;
  
  // Estado
  final String status;
  final bool isRequired;
  
  // Revisión médica de la foto
  final String? medicalReview;
  final DateTime? reviewedAt;
  final String? reviewedById;
  
  // Metadata
  final Map<String, dynamic>? metadata;
  final DateTime createdAt;
  final DateTime updatedAt;

  MedicalPhoto({
    required this.id,
    required this.certificateId,
    required this.userId,
    required this.requestedById,
    this.photoUrl,
    this.thumbnailUrl,
    this.originalFileName,
    this.fileSize,
    required this.bodyPart,
    this.bodyPartDetail,
    required this.photoType,
    this.description,
    required this.requestReason,
    this.requestInstructions,
    required this.requestDate,
    this.employeeNotes,
    this.photoDate,
    required this.status,
    required this.isRequired,
    this.medicalReview,
    this.reviewedAt,
    this.reviewedById,
    this.metadata,
    required this.createdAt,
    required this.updatedAt,
  });

  factory MedicalPhoto.fromJson(Map<String, dynamic> json) {
    return MedicalPhoto(
      id: json['id'],
      certificateId: json['certificateId'],
      userId: json['userId'],
      requestedById: json['requestedById'],
      photoUrl: json['photoUrl'],
      thumbnailUrl: json['thumbnailUrl'],
      originalFileName: json['originalFileName'],
      fileSize: json['fileSize'],
      bodyPart: json['bodyPart'],
      bodyPartDetail: json['bodyPartDetail'],
      photoType: json['photoType'],
      description: json['description'],
      requestReason: json['requestReason'],
      requestInstructions: json['requestInstructions'],
      requestDate: DateTime.parse(json['requestDate']),
      employeeNotes: json['employeeNotes'],
      photoDate: json['photoDate'] != null 
          ? DateTime.parse(json['photoDate'])
          : null,
      status: json['status'],
      isRequired: json['isRequired'] ?? true,
      medicalReview: json['medicalReview'],
      reviewedAt: json['reviewedAt'] != null 
          ? DateTime.parse(json['reviewedAt'])
          : null,
      reviewedById: json['reviewedById'],
      metadata: json['metadata'] != null 
          ? Map<String, dynamic>.from(json['metadata'])
          : null,
      createdAt: DateTime.parse(json['createdAt']),
      updatedAt: DateTime.parse(json['updatedAt']),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'certificateId': certificateId,
      'userId': userId,
      'requestedById': requestedById,
      'photoUrl': photoUrl,
      'thumbnailUrl': thumbnailUrl,
      'originalFileName': originalFileName,
      'fileSize': fileSize,
      'bodyPart': bodyPart,
      'bodyPartDetail': bodyPartDetail,
      'photoType': photoType,
      'description': description,
      'requestReason': requestReason,
      'requestInstructions': requestInstructions,
      'requestDate': requestDate.toIso8601String(),
      'employeeNotes': employeeNotes,
      'photoDate': photoDate?.toIso8601String(),
      'status': status,
      'isRequired': isRequired,
      'medicalReview': medicalReview,
      'reviewedAt': reviewedAt?.toIso8601String(),
      'reviewedById': reviewedById,
      'metadata': metadata,
      'createdAt': createdAt.toIso8601String(),
      'updatedAt': updatedAt.toIso8601String(),
    };
  }

  // Helper methods
  String get statusText {
    switch (status) {
      case 'requested':
        return 'Solicitada';
      case 'uploaded':
        return 'Subida';
      case 'reviewed':
        return 'Revisada';
      case 'rejected':
        return 'Rechazada';
      default:
        return 'Estado desconocido';
    }
  }

  String get photoTypeText {
    switch (photoType) {
      case 'injury':
        return 'Lesión';
      case 'lesion':
        return 'Lesión cutánea';
      case 'swelling':
        return 'Inflamación';
      case 'rash':
        return 'Erupción/Sarpullido';
      case 'wound':
        return 'Herida';
      case 'other':
        return 'Otro';
      default:
        return 'Tipo desconocido';
    }
  }

  bool get isRequested => status == 'requested';
  bool get isUploaded => status == 'uploaded';
  bool get isReviewed => status == 'reviewed';
  bool get isRejected => status == 'rejected';
  bool get isPending => isRequested || isUploaded;
  bool get hasPhoto => photoUrl != null;
}